### ISSUE TITLE
a bug found in save_model of LOMOTrainer

### ISSUE THREAD

**Issue by DingQiang2018** [OpenLMLab/LOMO#54]
我使用lomo微调chatglm2-6b，并使用LOMOTrainer的save_model方法保存。重新加载模型checkpoint后，我发现模型测出来的验证集loss与训练结束时测出来的不一样。我参考deepspeed官方保存模型的代码，重写了save_model（重写的代码如下），发现这个bug解决了。我建议作者考虑改写一下save_model。
I used LOMO to fine-tune chatglm2-6b and saved it using LOMOTrainer's save_model method. After reloading the model checkpoint, I found that the validation loss measured by the model differed from the loss measured at the end of training. I referred to the DeepSpeed official code, rewrote save_model (rewritten code below), and found this bug resolved. I suggest the authors rewriting save_model.
```
    def save_model(self, index):
        if self.training_args.local_rank in [-1, 0]:
            checkpoint_dir = sorted(Path(self.training_args.output_dir).glob("checkpoint-*"))
            if len(checkpoint_dir) >= self.training_args.save_total_limit:
                shutil.rmtree(checkpoint_dir[0], ignore_errors=True)
        torch.distributed.barrier()

        if self.training_args.resume_step:
            output_dir = os.path.join(self.training_args.output_dir, f"checkpoint-{index+self.training_args.resume_step}")
        else:
            output_dir = os.path.join(self.training_args.output_dir, f"checkpoint-{index}")
        if not os.path.exists(output_dir):
            os.makedirs(output_dir, exist_ok=True)

        state_dict = OrderedDict() if torch.distributed.get_rank() == 0 else None
        shared_params = {}

        # Prepare for checkpoint save by ensuring all parameters are partitioned
        self.model.optimizer.partition_all_parameters()

        with deepspeed.zero.GatheredParameters(list(self.model.module.parameters()), modifier_rank=0):
            if torch.distributed.get_rank() == 0:
                for name, param in self.model.module.named_parameters():
                    if param is None:
                        continue
                    # can't rely on param.data_ptr() as it will be reused as weights gets
                    # gathered and reduced, but param.ds_id is unique across all zero weights
                    # (and shared params will have the same param.ds_id)
                    if param.ds_id in shared_params:
                        # shared weights
                        #print(f"`{key}` is shared with `{shared_params[param.ds_id]}`")
                        state_dict[name] = state_dict[shared_params[param.ds_id]]
                    else:
                        state_dict[name] = param.detach().cpu()
                        shared_params[param.ds_id] = name
                    #print(f"param {param.ds_id} {param.shape} {key} ")

                # now buffers - not sure if need to take care of potentially shared weights here
                for name, buf in self.model.module.named_buffers():
                    if (buf is not None and name not in self.model.module._non_persistent_buffers_set):
                        state_dict[name] = buf.detach().cpu()

        if len(self.model.optimizer.persistent_parameters) > 0:
            self.model.optimizer.persistent_parameters[0].all_gather(self.model.optimizer.persistent_parameters)

        if torch.distributed.get_rank() == 0:
            torch.save(state_dict, os.path.join(output_dir, 'pytorch_model.bin'))

        torch.distributed.barrier()
```

**Comment by KaiLv69** [comment]
Thanks for your kindly feedback and `save_model()` has been updated according to your advice. FYI: https://github.com/OpenLMLab/LOMO/commit/06e50c07bb324be2863fe208012f8a9d6852b961

**Comment by yueg-security** [comment]
@DingQiang2018 您好，我注意到作者按照您的建议修改了LOMOTrainer和LOMOLoRaTrainer，LOMOTrainer运行没有问题，但LOMOLoRaTrainer会在self.model.optimizer.partition_all_parameters()处报错，您是否遇到了同样的问题呢？谢谢！

**Comment by wasifferoze** [comment]
> @DingQiang2018Hello, I noticed that the author modified LOMOTrainer and LOMOLoRaTrainer according to your suggestion. LOMOTrainer runs without problems, but LOMOLoRaTrainer will report an error at self.model.optimizer.partition_all_parameters(). Have you encountered the same problem? Thanks!

Yeah I am having this issue, did you find any solution?

**Comment by yueg-security** [comment]
> > @DingQiang2018Hello, I noticed that the author modified LOMOTrainer and LOMOLoRaTrainer according to your suggestion. LOMOTrainer runs without problems, but LOMOLoRaTrainer will report an error at self.model.optimizer.partition_all_parameters(). Have you encountered the same problem? Thanks!
> 
> Yeah I am having this issue, did you find any solution?
还没有解决……

**Comment by shawnricecake** [comment]
> > > @DingQiang2018Hello, I noticed that the author modified LOMOTrainer and LOMOLoRaTrainer according to your suggestion. LOMOTrainer runs without problems, but LOMOLoRaTrainer will report an error at self.model.optimizer.partition_all_parameters(). Have you encountered the same problem? Thanks!
> > 
> > 
> > Yeah I am having this issue, did you find any solution?
> > 还没有解决……

我也不能 在 merge llama with lora 之后得到相同的结果，很奇怪

**Comment by KaiLv69** [comment]
> @DingQiang2018 您好，我注意到作者按照您的建议修改了LOMOTrainer和LOMOLoRaTrainer，LOMOTrainer运行没有问题，但LOMOLoRaTrainer会在self.model.optimizer.partition_all_parameters()处报错，您是否遇到了同样的问题呢？谢谢！

Hi, lomo_lora_trainer中因为多了lora的optimizer所以不能通过model.optimizer来调用DeepSpeedZeRoOffload。我目前把lomo_lora_trainer.py中的save_model()回退到之前版本了。

**Comment by KaiLv69** [comment]
> 很荣幸我的建议被采纳。我还想问问您对之前`save_model`出错的具体原因有什么看法吗？我还没想通。 It's my pleasure to see my advice being accepted. Moreover, may I ask you if you have any comment on why `save_model` went wrong previously? I have not figured out this.

Hi, 我想知道使用ChatGLM2的loss两种保存方法会查多少，不知道您是否还有记录？BTW，LLaMA会有同样的问题吗？

**Comment by shawnricecake** [comment]
> > @DingQiang2018 您好，我注意到作者按照您的建议修改了LOMOTrainer和LOMOLoRaTrainer，LOMOTrainer运行没有问题，但LOMOLoRaTrainer会在self.model.optimizer.partition_all_parameters()处报错，您是否遇到了同样的问题呢？谢谢！
> 
> Hi, lomo_lora_trainer中因为多了lora的optimizer所以不能通过model.optimizer来调用DeepSpeedZeRoOffload。我目前把lomo_lora_trainer.py中的save_model()回退到之前版本了。

Hi, 我注意到了，但是我目前还是没办法做到 merge 之后的 model 有相同的eval resutls。。。。
